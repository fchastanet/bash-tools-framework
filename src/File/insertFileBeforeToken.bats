#!/usr/bin/env bash

# shellcheck source=src/batsHeaders.sh
source "$(cd "${BATS_TEST_DIRNAME}/.." && pwd)/batsHeaders.sh"

# shellcheck source=/src/File/insertFileBeforeToken.sh
source "${srcDir}/File/insertFileBeforeToken.sh"

function File::insertFileBeforeToken::OneToken { #@test
  cp "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken.txt" \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken.txt"
  File::insertFileBeforeToken \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken.toInject.txt" \
    "# FUNCTIONS"

  diff "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken.expected.txt" \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken.txt"
}

function File::insertFileBeforeToken::TwoTokens { #@test
  cp "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken2.txt" \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken2.txt"
  File::insertFileBeforeToken \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken2.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken.toInject.txt" \
    "# FUNCTIONS"

  diff "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken2.expected.txt" \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken2.txt"
}

function File::insertFileBeforeToken::InsertTwice { #@test
  cp "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken2.txt" \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken3.txt"
  File::insertFileBeforeToken \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken3.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken.toInject.txt" \
    "# FUNCTIONS"
  File::insertFileBeforeToken \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken3.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken.toInject3.txt" \
    "# FUNCTIONS"

  diff "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken3.expected.txt" \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken3.txt"
}

function File::insertFileBeforeToken::NoMatchingPattern { #@test
  cp "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken2.txt" \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken4.txt"
  run File::insertFileBeforeToken \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken4.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken.toInject.txt" \
    "# NOT MATCHING PATTERN"
  assert_success
  diff "${BATS_TEST_DIRNAME}/testsData/insertFileBeforeToken2.txt" \
    "${BATS_RUN_TMPDIR}/insertFileBeforeToken4.txt"
}
