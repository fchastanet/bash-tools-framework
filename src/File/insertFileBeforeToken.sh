#!/usr/bin/env bash

# Public: insert file content inside another file before a
# pattern
#
# @param {String} file $1 file in which fileToImport will be
#   injected
# @param {String} fileToImport $2 file to inject before the
#   token
# @param {String} token $3 token needs to be properly escaped
#   for usage inside sed // regexp
File::insertFileBeforeToken() {
  local file="$1"
  local fileToImport="$2"
  local token="$3"

  # old version inserts at each occurrentce
  # sed -E -i "0,/${token}/ {
  #     h
  #     r ${fileToImport}
  #     g
  #     N
  #   }" "${file}"

  sed -i -E -e "
      /${token}/!b
      r ${fileToImport}
      N;:a;n;$!ba
    " "${file}"
}
