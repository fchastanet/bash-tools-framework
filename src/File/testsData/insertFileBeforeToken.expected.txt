stuff before
insertFileBeforeToken injected
# FUNCTIONS
stuff after
